File: frontend/src/types/listing.ts
export type UserPublic = {
  id: number
  telegram_id: number
  username: string | null
  first_name: string
  last_name: string | null
  photo_url: string | null
}

export type PhotoResponse = {
  id: number
  photoUrl: string
  displayOrder: number
  thumbnailData: string | null  // Base64-encoded thumbnail
  fileSizeBytes: number | null
  originalFilename: string | null
  createdAt: string
}

export type PhotoUploadResponse = {
  url: string
  thumbnail: string | null  // Base64-encoded thumbnail
}

export type Listing = {
  id: number
  title: string
  description: string
  priceMinorUnits: number
  currency: string
  sellerId: number
  createdAt: string
  updatedAt: string
  photos: PhotoResponse[]
  photoUrl: string | null  // Deprecated: use photos[0]?.photoUrl instead
  seller: UserPublic
}

export type ListingViewModel = Listing & {
  priceLabel: string
}

File: backend/app/schemas/listing.py
from datetime import datetime

from pydantic import BaseModel, Field, PositiveInt

from app.schemas.photo import PhotoResponse
from app.schemas.user import UserPublic


class ListingBase(BaseModel):
    title: str = Field(..., max_length=120)
    description: str = Field(..., max_length=2048)
    price_minor_units: PositiveInt
    currency: str = Field(..., min_length=3, max_length=3)


class ListingCreate(ListingBase):
    category: str | None = Field(None, max_length=50)
    condition: str | None = Field(None, max_length=20)


class Listing(ListingBase):
    id: int
    seller_id: int
    created_at: datetime
    updated_at: datetime
    photos: list[PhotoResponse] = []

    class Config:
        from_attributes = True


class ListingWithSeller(Listing):

    seller: UserPublic

    class Config:
        from_attributes = True
